src/analysis/processing/qgsalgorithmkmeansclustering.cpp

```diff
@@ -0,0 +1,371 @@
+        distances[j] = std::min( points[j].point.sqrDist( centers[i - 1] ), distances[j] );
+
+        // greedily take a point that's farthest from any of accepted clusters
+        if ( distances[j] > maxDistance )
+        {
+          candidateCenter = j;
+          maxDistance = distances[j];
+        }
+      }
+
+      // checked earlier by counting entries on input, just in case
+      Q_ASSERT( maxDistance >= 0 );
+
+      // accept candidate to centers
+      distances[candidateCenter] = -1;
+      // copy the point coordinates into the initial centers array
+      centers[i] = points[candidateCenter].point;
+    }
+  }
+}
+
+// ported from https://github.com/postgis/postgis/blob/svn-trunk/liblwgeom/lwkmeans.c
+
+void QgsKMeansClusteringAlgorithm::calculateKMeans( std::vector<QgsKMeansClusteringAlgorithm::Feature> &objs, std::vector<QgsPointXY> &centers, int k, QgsProcessingFeedback *feedback )
+{
+  int converged = false;
+  bool changed = false;
+
+  // avoid reallocating weights array for every iteration
+  std::vector< uint > weights( k );
+
+  uint i = 0;
+  for ( i = 0; i < KMEANS_MAX_ITERATIONS && !converged; i++ )
+  {
+    if ( feedback && feedback->isCanceled() )
+      break;
+
+    findNearest( objs, centers, k, changed );
+    updateMeans( objs, centers, weights, k );
+    converged = !changed;
+  }
+
+  if ( !converged && feedback )
+    feedback->reportError( QObject::tr( "Clustering did not converge after %1 iterations" ).arg( i ) );
+  else if ( feedback )
+    feedback->pushInfo( QObject::tr( "Clustering converged after %1 iterations" ).arg( i ) );
+}
+
+// ported from https://github.com/postgis/postgis/blob/svn-trunk/liblwgeom/lwkmeans.c
+
+void QgsKMeansClusteringAlgorithm::findNearest( std::vector<QgsKMeansClusteringAlgorithm::Feature> &points, const std::vector<QgsPointXY> &centers, const int k, bool &changed )
+{
+  changed = false;
+  ulong n = points.size();
+  for ( ulong i = 0; i < n; i++ )
+  {
+    Feature &point = points[i];
+
+    // Initialize with distance to first cluster
+    double currentDistance = point.point.sqrDist( centers[0] );
+    int currentCluster = 0;
+
+    // Check all other cluster centers and find the nearest
+    for ( int cluster = 1; cluster < k; cluster++ )
+    {
+      const double distance = point.point.sqrDist( centers[cluster] );
+      if ( distance < currentDistance )
+      {
+        currentDistance = distance;
+        currentCluster = cluster;
+      }
+    }
+
+    // Store the nearest cluster this object is in
+    if ( point.cluster != currentCluster )
+    {
+      changed = true;
+      point.cluster = currentCluster;
+    }
+  }
+}
+
+// ported from https://github.com/postgis/postgis/blob/svn-trunk/liblwgeom/lwkmeans.c
+
+void QgsKMeansClusteringAlgorithm::updateMeans( const std::vector<Feature> &points, std::vector<QgsPointXY> &centers, std::vector<uint> &weights, const int k )
+{
+  uint n = points.size();
```

Comment:
Previously in `findNearest` or `initClusters`, you used a `ulong` for the size of the `std::vector` container. What do you think about using `std::size_t` instead for homogeneity?